File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(cur: TreeNode):
            if cur is None:
                return True, -1

            lhs_balanced, lhs_height = dfs(cur.left)
            rhs_balanced, rhs_height = dfs(cur.right)
            tree_height = max(lhs_height, rhs_height) + 1
            
            if lhs_balanced and rhs_balanced and abs(lhs_height - rhs_height) <= 1:
                return True, tree_height

            return False, tree_height

        is_balanced, height = dfs(root)
        return is_balanced

        

File: Data Structures & Algorithms/max-water-container/submission-2.go
func maxArea(height []int) int {
    l := 0
    r := len(height) - 1
    maxArea := 0

    for l < r {
        maxArea = max(maxArea, (r - l) * min(height[l], height[r]))
        if height[l] < height[r] {
            l++
        } else {
            r--
        }
    }
    return maxArea
}